CSV importer enhancement to enable to users to upload a consolidated …

namitgupta opened the pull request:
Hi @justinedelson @davidjgonzalez ,

Hi,

At present the CSV importer doesn't support bulk asset upload from a system other than the one on which AEM instance is running.

We have been receiving many requests from the customers for enabling their respective business teams to do bulk upload of images in an archive format along with the metadata.

We have made some enhancements to the CSV importer tool to incorporate the aforesaid feature.

Kindly allow me to commit the changes.

Regards,
Namit

namigupt committed:
CSV importer enhancement to enable to users to upload a consolidated zip file with assets and metadata.

justinedelson commented:
Please do not make unrelated formatting changes. That makes it very difficult to see what is actually being changed as part of this request.

bq. At present the CSV importer doesn't support bulk asset upload from a system other than the one on which AEM instance is running.

FWIW, this statement is incorrect. The only requirement is that the files must be accessible from the AEM system. This could be on a shared file server. Theoretically, the CSV importer could also be enhanced to support, for example, FTP or S3 URLs.

justinedelson reviewed (changes requested):
please resubmit without formatting changes.

namigupt committed:
CSV importer enhancement to enable to users to upload a consolidated zip file with assets and metadata. | Removed formatting changes.

namigupt committed:
CSV importer enhancement to enable to users to upload a consolidated zip file with assets and metadata. | Removed formatting changes.

justinedelson commented on bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/CsvAssetImporterServlet.java:
```diff
@@ -89,8 +89,30 @@ protected final void doPost(final SlingHttpServletRequest request, final SlingHt
+        	Iterator<String[]> rows = null;
+			final long start = System.currentTimeMillis();
+			String inFileMimeType = params.getInFileMimeType();
+			final String tempDir = System.getProperty("java.io.tmpdir") + start;
```
I'm not sure that `java.io.tmpdir` is guaranteed to end in a slash. It seems like it would be better to do something like:

     final File tmpDir = new File(FileUtils.getTempDirectory(), "acs-tools-csv-importer-" + start);
     tmpDir.mkdir();

justinedelson replied:
I'm not sure that `java.io.tmpdir` is guaranteed to end in a slash. It seems like it would be better to do something like:

     final File tmpDir = new File(FileUtils.getTempDirectory(), "acs-tools-csv-importer-" + start);
     tmpDir.mkdir();

justinedelson replied:
Also, as far as I can see, the temp directory is never removed. That's very problematic.

justinedelson commented on bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/CsvAssetImporterServlet.java:
```diff
@@ -89,8 +89,30 @@ protected final void doPost(final SlingHttpServletRequest request, final SlingHt
-            final long start = System.currentTimeMillis();
-            final Iterator<String[]> rows = this.getRowsFromCsv(params);
+        	Iterator<String[]> rows = null;
+			final long start = System.currentTimeMillis();
```
It's pretty clear you are mixing tabs and spaces here. Please don't do that and respect the original formatting.

namitgupta commented:
@justinedelson 

Apologies! 

I have reverted the formatting changes.

Please review now.

justinedelson commented on bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/Parameters.java:
```diff
@@ -289,4 +293,18 @@ public String toString() {
+	/**
+	 * @return the inFileMimeType
+	 */
+	public String getInFileMimeType() {
```
not sure why this isn't just called `getMimeType()` but maybe there's a reason I'm missing.

justinedelson commented on bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/UnzipUtility.java:
```diff
@@ -0,0 +1,103 @@
+package com.adobe.acs.tools.csv_asset_importer.impl;
```
missing license header

justinedelson commented on bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/UnzipUtility.java:
```diff
@@ -0,0 +1,103 @@
+	 * @param destDirectory
+	 * @throws IOException
+	 */
+	public static void unzip(String zipFilePath, String destDirectory) throws IOException {
```
AFAIK, this method is never used and is largely duplicative of the next method. Please remove it.

justinedelson commented on bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/UnzipUtility.java:
```diff
@@ -0,0 +1,103 @@
+	 * @param filePath
+	 * @throws IOException
+	 */
+	private static void extractFile(ZipInputStream zipIn, String filePath) throws IOException {
```
This method isn't necessary. Please use `FileUtils.copyInputStreamToFile()` or equivalent.

badvision replied:
Rather than extracting the files and importing them, which creates a lot of IO and uses a lot of temp space, why not just read the zip contents in a stream and work off the stream?  You can take a zip file input stream and pass that directly in to the asset API to create assets.  The result would be zip contents directly ingested with no middle steps.  This would greatly improve throughput and reduce resource usage.

justinedelson commented on bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/CsvAssetImporterServlet.java:
```diff
@@ -89,8 +89,30 @@ protected final void doPost(final SlingHttpServletRequest request, final SlingHt
+				InputStream inputStream = params.getFile();
+
+				// un-archive the zip file.
+				UnzipUtility.unzip(inputStream, tempDir);
```
Instead of having this code in the project, what about embedding https://github.com/zeroturnaround/zt-zip ?

namitgupta commented:
i tested this with open jdk n oracle jdk on windows, mac n centOS.

however, i can incorporate the suggested changes in case there are specific
failure scenarios.

On May 9, 2017 8:36 PM, "justinedelson" <notifications@github.com> wrote:

*@justinedelson* commented on this pull request.
------------------------------

In bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/
CsvAssetImporterServlet.java
<https://github.com/Adobe-Consulting-Services/acs-aem-tools/pull/189#discussion_r115514299>
:

> @@ -89,8 +89,30 @@ protected final void doPost(final SlingHttpServletRequest request, final SlingHt
         final Parameters params = new Parameters(request);
         if (params.getFile() != null) {

-            final long start = System.currentTimeMillis();
-            final Iterator<String[]> rows = this.getRowsFromCsv(params);
+        	Iterator<String[]> rows = null;
+			final long start = System.currentTimeMillis();
+			String inFileMimeType = params.getInFileMimeType();
+			final String tempDir = System.getProperty("java.io.tmpdir") + start;

I'm not sure that java.io.tmpdir is guaranteed to end in a slash. It seems
like it would be better to do something like:

 final File tmpDir = new File(FileUtils.getTempDirectory(),
"acs-tools-csv-importer-" + start);
 tmpDir.mkdir();

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Adobe-Consulting-Services/acs-aem-tools/pull/189#pullrequestreview-37058250>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ACClss7CeK-7ORClPhXhZ34hUOSBbMLLks5r4ID3gaJpZM4NVWVM>
.

justinedelson commented on bundle/src/main/java/com/adobe/acs/tools/csv_asset_importer/impl/CsvAssetImporterServlet.java:
```diff
@@ -89,8 +89,30 @@ protected final void doPost(final SlingHttpServletRequest request, final SlingHt
+
+				IOUtils.closeQuietly(inputStream);
+
+				InputStream is = new FileInputStream(tempDir + "/import.csv");
```
Would it make more sense to look for the first file with a csv extension?

justinedelson reviewed (changes requested):
In addition the comments I've made, I suspect some UI change is necessary in order to make sure users know they can upload a ZIP file and how the CSV file is pulled from the ZIP.

namigupt committed:
Incorporated review comments:

1. java.io.tmpdir is not guaranteed to end with slash: fixed.

2. Formatting issue: Fixed

3. Temp files not removed: fixed.

4. Rename getInFileMimeType() : fixed.

5.missing license header in UnzipUtility.java: fixed.

6. remove boilerplate code from unzipUtility.java: fixed.

7. pick first file in the zip as the import seed file: I have not fixed this as the idea behind this is to impose a standard zip structure.

8. embed https://github.com/zeroturnaround/zt-zip for unzipping function: This library has got a lot of code which is not needed in our case. Therefore, we should stick with the UnzipUtility.java.

namigupt committed:
Removed unnecessary boilerplate code.

namitgupta commented:
Hi @justinedelson 

Incorporated review comments:

1. java.io.tmpdir is not guaranteed to end with slash: fixed.

2. Formatting issue: Fixed

3. Temp files not removed: fixed.

4. Rename getInFileMimeType() : fixed.

5.missing license header in UnzipUtility.java: fixed.

6. remove boilerplate code from unzipUtility.java: fixed.

7. pick first file in the zip as the import seed file: I have not fixed this as the idea behind this is to impose a standard zip structure.

8. embed https://github.com/zeroturnaround/zt-zip for unzipping function: This library has got a lot of code which is not needed in our case. Therefore, we should stick with the UnzipUtility.java.

kdesineedi commented:
@justinedelson May not be related to this PR but I have a general question as to if this tool is supported for AEMAsACS or we should use the "https://experienceleague.adobe.com/en/docs/experience-manager-cloud-service/content/assets/assets-view/bulk-import-assets-view" instead? Please can you provide some clarity or guidance on what is the difference between the both?

davidjgonzalez commented:
Hi @kdesineedi  - Bulk Import  is a productized way to import asset for AEM CS. This feature NOT supported by AEM CS since this streams binary data via the JVM which is an anti-pattern. AEM Bulk Import imports assets in a supported manner, and as a bonus, is supported by Adobe Support.
